certmon/monitord/impl.go

```diff
@@ -0,0 +1,238 @@
+package monitord
+
+import (
+	"bufio"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/Symantec/Dominator/lib/log"
+	"github.com/Symantec/Dominator/lib/log/prefixlogger"
+	"github.com/Symantec/keymaster/proto/certmon"
+	"golang.org/x/crypto/ssh"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const bufferLength = 16
+
+type receiveType struct {
+	certType uint32
+	certData []byte
+}
+
+func newMonitor(keymasterServerHostname string, keymasterServerPortNum uint,
+	logger log.Logger) (*Monitor, error) {
+	sshRawCertChannel := make(chan []byte, bufferLength)
+	sshCertChannel := make(chan *ssh.Certificate, bufferLength)
+	x509RawCertChannel := make(chan []byte, bufferLength)
+	x509CertChannel := make(chan *x509.Certificate, bufferLength)
+	monitor := &Monitor{
+		keymasterServerHostname: keymasterServerHostname,
+		keymasterServerPortNum:  keymasterServerPortNum,
+		closers:                 make(map[string]chan<- struct{}),
+		// Transmit side channels (private).
+		sshRawCertChannel:  sshRawCertChannel,
+		sshCertChannel:     sshCertChannel,
+		x509RawCertChannel: x509RawCertChannel,
+		x509CertChannel:    x509CertChannel,
+		// Receive side channels (public).
+		SshRawCertChannel:  sshRawCertChannel,
+		SshCertChannel:     sshCertChannel,
+		X509RawCertChannel: x509RawCertChannel,
+		X509CertChannel:    x509CertChannel,
+	}
+	go monitor.forever(logger)
+	return monitor, nil
+}
+
+func (m *Monitor) forever(logger log.Logger) {
```

Comment:
forever.... and ever is a great chorus.... but not very descriptive... a better name would be nice... again just mark it a as a TODO.

Reply:
OK, I've renamed to `monitorForever()`.